activesupport/test/message_encryptor_test.rb

```diff
@@ -132,3 +140,175 @@ def munge(base64_string)
+    encrypted_message = @encryptor.encrypt_and_sign(string_message, expires_at: 1.hour.from_now)
+
+    travel 59.minutes
+    assert_equal string_message, @encryptor.decrypt_and_verify(encrypted_message)
+
+    travel 2.minutes
+    assert_nil @encryptor.decrypt_and_verify(encrypted_message)
+  end
+
+  def test_set_relative_expiration_date_by_passing_expires_in
+    encrypted_message = encryptor.encrypt_and_sign(string_message, expires_in: 2.hours)
+
+    travel 1.hour
+    assert_equal string_message, encryptor.decrypt_and_verify(encrypted_message)
+
+    travel 1.hour + 1.second
+    assert_nil encryptor.decrypt_and_verify(encrypted_message)
+  end
+
+  def test_passing_expires_in_less_than_a_second_is_not_expired
+    freeze_time do
+      encrypted_message = @encryptor.encrypt_and_sign(string_message, expires_in: 1.second)
+      
+      travel 0.5.seconds
+      assert_equal string_message, @encryptor.decrypt_and_verify(encrypted_message)
+
+      travel 1.second
+      assert_nil @encryptor.decrypt_and_verify(encrypted_message)
+    end
+  end
+
+  def test_favor_expires_at_over_expires_in
+    payment_related_message = @encryptor.encrypt_and_sign(string_message, purpose: "payment", expires_at: 2.year.from_now, expires_in: 1.second)
+
+    travel 1.year
+    assert_equal string_message, @encryptor.decrypt_and_verify(payment_related_message, purpose: :payment) 
+
+    travel 1.year + 1.day
+    assert_nil @encryptor.decrypt_and_verify(payment_related_message, purpose: "payment")
+  end
+
+  def test_skip_expires_at_and_expires_in_to_disable_expiration_check
+    payment_related_message = @encryptor.encrypt_and_sign(string_message, purpose: "payment")
+
+    travel 100.years
+    assert_equal string_message, @encryptor.decrypt_and_verify(payment_related_message, purpose: "payment")
+  end
+=begin
+  def test_hash_messages
+    json_encryptor = ActiveSupport::MessageEncryptor.new(@secret, serializer: MessageEncryptorTest::JSONSerializer.new)
+
+    no_purpose = json_encryptor.encrypt_and_sign(hash_message)
+    assert_equal hash_message.stringify_keys, json_encryptor.decrypt_and_verify(no_purpose)
+
+    payment_related_message = json_encryptor.encrypt_and_sign(hash_message, purpose: "payment", expires_at: 1.year.from_now)
+    assert_equal hash_message.stringify_keys, json_encryptor.decrypt_and_verify(payment_related_message, purpose: "payment")
+
+    marshal_encryptor = ActiveSupport::MessageEncryptor.new(@secret, serializer: Marshal)
+
+    no_purpose = marshal_encryptor.encrypt_and_sign(hash_message)
+    assert_equal hash_message, marshal_encryptor.decrypt_and_verify(no_purpose)
+
+    payment_related_message = marshal_encryptor.encrypt_and_sign(hash_message, purpose: "payment", expires_at: 1.year.from_now)
+    assert_equal hash_message, marshal_encryptor.decrypt_and_verify(payment_related_message, purpose: :payment)
+    assert_nil marshal_encryptor.decrypt_and_verify(payment_related_message, purpose: :login)
+  end
+
```

Comment:
If the keys deserialized aren't left as strings wouldn't backward compatibility be affected? (JSON)